(Employee Performance Analysis)
 INX Future inc.

 1. Department wise performances. 
 2. Top Important Factors effecting employee performance.
 3. A trained model which can predict the employee performance based on factors as inputs.
    This will be used to hire employees.
 4. Recommendations to improve the employee performance based on insights from analysis. 

In [1]:
import pandas as pd
import numpy as np

# load data

In [2]:
#Rename columns as q: continuous,c: catgorical, y: outcome
names = ['c_EmpNumber','q_Age','c_Gender','c_EducationBackground','c_MaritalStatus','c_EmpDepartment','c_EmpJobRole','c_BusinessTravelFrequency',
         'q_DistanceFromHome','q_EmpEducationLevel','q_EmpEnvironmentSatisfaction','q_EmpHourlyRate','q_EmpJobInvolvement','q_EmpJobLevel'
        ,'q_EmpJobSatisfaction','q_NumCompaniesWorked','c_OverTime','q_EmpLastSalaryHikePercent','q_EmpRelationshipSatisfaction','q_TotalWorkExperienceInYears',
        'q_TrainingTimesLastYear','q_EmpWorkLifeBalance','q_ExperienceYearsAtThisCompany','q_ExperienceYearsInCurrentRole','q_YearsSinceLastPromotion',
        'q_YearsWithCurrManager','c_Attrition','y_PerformanceRating']

In [5]:
df1 = pd.read_excel("INX_Future_Inc_Employee_Performance_CDS_Project2_Data_V1.8.xls",names=names)
df1.head(5)

c_EmpNumber  q_Age c_Gender c_EducationBackground c_MaritalStatus  \
0    E1001000     32     Male             Marketing          Single   
1    E1001006     47     Male             Marketing          Single   
2    E1001007     40     Male         Life Sciences         Married   
3    E1001009     41     Male       Human Resources        Divorced   
4    E1001010     60     Male             Marketing          Single   

   c_EmpDepartment     c_EmpJobRole c_BusinessTravelFrequency  \
0            Sales  Sales Executive             Travel_Rarely   
1            Sales  Sales Executive             Travel_Rarely   
2            Sales  Sales Executive         Travel_Frequently   
3  Human Resources          Manager             Travel_Rarely   
4            Sales  Sales Executive             Travel_Rarely   

   q_DistanceFromHome  q_EmpEducationLevel  ...  \
0                  10                    3  ...   
1                  14                    4  ...   
2                   5                    4  ...   
3                  10                    4  ...   
4                  16                    4  ...   

   q_EmpRelationshipSatisfaction  q_TotalWorkExperienceInYears  \
0                              4                            10   
1                              4                            20   
2                              3                            20   
3                              2                            23   
4                              4                            10   

   q_TrainingTimesLastYear  q_EmpWorkLifeBalance  \
0                        2                     2   
1                        2                     3   
2                        2                     3   
3                        2                     2   
4                        1                     3   

   q_ExperienceYearsAtThisCompany  q_ExperienceYearsInCurrentRole  \
0                              10                               7   
1                               7                               7   
2                              18                              13   
3                              21                               6   
4                               2                               2   

  q_YearsSinceLastPromotion  q_YearsWithCurrManager  c_Attrition  \
0                         0                       8           No   
1                         1                       7           No   
2                         1                      12           No   
3                        12                       6           No   
4                         2                       2           No   

   y_PerformanceRating  
0                    3  
1                    3  
2                    4  
3                    3  
4                    3  

[5 rows x 28 columns]

# Data Sanity checks - Finding, deleting or updating the incorrigible cells/rows

In [6]:
df1.shape

(1200, 28)

In [7]:
# All categorical show as objects as they are categorical string all other fields show as integer
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1200 entries, 0 to 1199
Data columns (total 28 columns):
 #   Column                          Non-Null Count  Dtype 
---  ------                          --------------  ----- 
 0   c_EmpNumber                     1200 non-null   object
 1   q_Age                           1200 non-null   int64 
 2   c_Gender                        1200 non-null   object
 3   c_EducationBackground           1200 non-null   object
 4   c_MaritalStatus                 1200 non-null   object
 5   c_EmpDepartment                 1200 non-null   object
 6   c_EmpJobRole                    1200 non-null   object
 7   c_BusinessTravelFrequency       1200 non-null   object
 8   q_DistanceFromHome              1200 non-null   int64 
 9   q_EmpEducationLevel             1200 non-null   int64 
 10  q_EmpEnvironmentSatisfaction    1200 non-null   int64 
 11  q_EmpHourlyRate                 1200 non-null   int64 
 12  q_EmpJobInvolvement             1200 non-null   

In [8]:
#Check Isnull and isna - no blanks or nans
print(df1.isna().sum(), df1.isnull().sum())

c_EmpNumber                       0
q_Age                             0
c_Gender                          0
c_EducationBackground             0
c_MaritalStatus                   0
c_EmpDepartment                   0
c_EmpJobRole                      0
c_BusinessTravelFrequency         0
q_DistanceFromHome                0
q_EmpEducationLevel               0
q_EmpEnvironmentSatisfaction      0
q_EmpHourlyRate                   0
q_EmpJobInvolvement               0
q_EmpJobLevel                     0
q_EmpJobSatisfaction              0
q_NumCompaniesWorked              0
c_OverTime                        0
q_EmpLastSalaryHikePercent        0
q_EmpRelationshipSatisfaction     0
q_TotalWorkExperienceInYears      0
q_TrainingTimesLastYear           0
q_EmpWorkLifeBalance              0
q_ExperienceYearsAtThisCompany    0
q_ExperienceYearsInCurrentRole    0
q_YearsSinceLastPromotion         0
q_YearsWithCurrManager            0
c_Attrition                       0
y_PerformanceRating         

In [9]:
#Perform Basic Consistency checks
from collections import Counter
Counter(df1.q_Age)  # No problem             
df1.q_Age.sum()                                

44302

In [10]:
Counter(df1.c_Gender) 

Counter({'Male': 725, 'Female': 475})

In [11]:
df1.q_EmpEducationLevel.sum()

3471

In [12]:
Counter(df1.c_EducationBackground) # No Problem 

Counter({'Marketing': 137,
         'Life Sciences': 492,
         'Human Resources': 21,
         'Medical': 384,
         'Other': 66,
         'Technical Degree': 100})

In [13]:
Counter(df1.c_EmpDepartment)  # No Problem 

Counter({'Sales': 373,
         'Human Resources': 54,
         'Development': 361,
         'Data Science': 20,
         'Research & Development': 343,
         'Finance': 49})

In [14]:
Counter(df1.c_MaritalStatus) # No Problem 

Counter({'Single': 384, 'Married': 548, 'Divorced': 268})

In [15]:
Counter(df1.c_EmpJobRole)

Counter({'Sales Executive': 270,
         'Manager': 51,
         'Developer': 236,
         'Sales Representative': 69,
         'Human Resources': 45,
         'Senior Developer': 52,
         'Data Scientist': 20,
         'Senior Manager R&D': 15,
         'Laboratory Technician': 64,
         'Manufacturing Director': 33,
         'Research Scientist': 77,
         'Healthcare Representative': 33,
         'Research Director': 19,
         'Manager R&D': 94,
         'Finance Manager': 49,
         'Technical Architect': 7,
         'Business Analyst': 16,
         'Technical Lead': 38,
         'Delivery Manager': 12})

In [16]:
Counter(df1.c_BusinessTravelFrequency)  # No problem 

Counter({'Travel_Rarely': 846, 'Travel_Frequently': 222, 'Non-Travel': 132})

In [17]:
Counter(df1.c_Gender)         # No problem  

Counter({'Male': 725, 'Female': 475})

In [18]:
df1.q_EmpJobLevel.sum()

2481

In [19]:
df1.q_EmpJobSatisfaction.sum()

3279

In [20]:
df1.q_EmpHourlyRate.sum()

79178

In [21]:
Counter(df1.c_OverTime)

Counter({'No': 847, 'Yes': 353})

In [22]:
Counter(df1.y_PerformanceRating)

Counter({3: 874, 4: 132, 2: 194})

In [23]:
#Check Data Constency for every column. Note continues need not be checked as sum in above step is enoug
# first Establish a single Categorical Columns which has no discrepencies and every cross tab ensure the total as 32537
pd.crosstab(df1.c_EmpJobRole,df1.y_PerformanceRating,margins='all')# work class has an issue

y_PerformanceRating          2    3    4   All
c_EmpJobRole                                  
Business Analyst             0   13    3    16
Data Scientist               1   17    2    20
Delivery Manager             0   12    0    12
Developer                    6  199   31   236
Finance Manager             15   30    4    49
Healthcare Representative    8   22    3    33
Human Resources              9   31    5    45
Laboratory Technician       14   45    5    64
Manager                     12   29   10    51
Manager R&D                 17   66   11    94
Manufacturing Director       3   28    2    33
Research Director            5   11    3    19
Research Scientist          17   46   14    77
Sales Executive             64  181   25   270
Sales Representative        12   54    3    69
Senior Developer             6   40    6    52
Senior Manager R&D           4   10    1    15
Technical Architect          0    7    0     7
Technical Lead               1   33    4    38
All                        194  874  132  1200

In [24]:
pd.crosstab(df1.c_EducationBackground,df1.y_PerformanceRating,margins='all')# np

y_PerformanceRating      2    3    4   All
c_EducationBackground                     
Human Resources          3   16    2    21
Life Sciences           78  357   57   492
Marketing               29   94   14   137
Medical                 63  281   40   384
Other                    3   56    7    66
Technical Degree        18   70   12   100
All                    194  874  132  1200

In [25]:
pd.crosstab(df1.c_EmpDepartment,df1.y_PerformanceRating,margins='all')# np

y_PerformanceRating       2    3    4   All
c_EmpDepartment                            
Data Science              1   17    2    20
Development              13  304   44   361
Finance                  15   30    4    49
Human Resources          10   38    6    54
Research & Development   68  234   41   343
Sales                    87  251   35   373
All                     194  874  132  1200

In [26]:
pd.crosstab(df1.c_MaritalStatus,df1.y_PerformanceRating,margins='all')# np

y_PerformanceRating    2    3    4   All
c_MaritalStatus                         
Divorced              37  205   26   268
Married              100  393   55   548
Single                57  276   51   384
All                  194  874  132  1200

In [27]:
pd.crosstab(df1.c_Gender,df1.y_PerformanceRating,margins='all')

y_PerformanceRating    2    3    4   All
c_Gender                                
Female                75  349   51   475
Male                 119  525   81   725
All                  194  874  132  1200

In [28]:
pd.crosstab(df1.c_OverTime,df1.y_PerformanceRating,margins='all')# np

y_PerformanceRating    2    3    4   All
c_OverTime                              
No                   155  595   97   847
Yes                   39  279   35   353
All                  194  874  132  1200

In [29]:
pd.crosstab(df1.c_BusinessTravelFrequency,df1.y_PerformanceRating,margins='all')# np

y_PerformanceRating          2    3    4   All
c_BusinessTravelFrequency                     
Non-Travel                  21   93   18   132
Travel_Frequently           37  154   31   222
Travel_Rarely              136  627   83   846
All                        194  874  132  1200

In [30]:
pd.crosstab(df1.q_Age,df1.y_PerformanceRating,margins='all') #np

y_PerformanceRating    2    3    4   All
q_Age                                   
18                     0    7    1     8
19                     0    4    4     8
20                     0    6    0     6
21                     3    7    1    11
22                     2   12    1    15
23                     1    6    2     9
24                     6   12    2    20
25                     4   18    2    24
26                     4   24    5    33
27                     6   34    3    43
28                     2   33    4    39
29                    11   35    5    51
30                     5   30    7    42
31                    12   37    8    57
32                     3   39    4    46
33                    10   32    4    46
34                    12   54    5    71
35                    10   51    3    64
36                    10   41    9    60
37                     5   30    6    41
38                     4   38    6    48
39                     8   19    4    31
40                     9   30    7    46
41                     5   23    4    32
42                     2   29    5    36
43                     6   17    3    26
44                     7   22    1    30
45                     5   25    6    36
46                     6   16    2    24
47                     4   15    1    20
48                     3   12    1    16
49                     3   17    1    21
50                     3   22    3    28
51                     1   11    2    14
52                     4    9    2    15
53                     4   11    0    15
54                     3   13    0    16
55                     3   12    2    17
56                     2    7    2    11
57                     1    3    0     4
58                     3    6    2    11
59                     2    3    1     6
60                     0    2    1     3
All                  194  874  132  1200

In [31]:
df1.columns

Index(['c_EmpNumber', 'q_Age', 'c_Gender', 'c_EducationBackground',
       'c_MaritalStatus', 'c_EmpDepartment', 'c_EmpJobRole',
       'c_BusinessTravelFrequency', 'q_DistanceFromHome',
       'q_EmpEducationLevel', 'q_EmpEnvironmentSatisfaction',
       'q_EmpHourlyRate', 'q_EmpJobInvolvement', 'q_EmpJobLevel',
       'q_EmpJobSatisfaction', 'q_NumCompaniesWorked', 'c_OverTime',
       'q_EmpLastSalaryHikePercent', 'q_EmpRelationshipSatisfaction',
       'q_TotalWorkExperienceInYears', 'q_TrainingTimesLastYear',
       'q_EmpWorkLifeBalance', 'q_ExperienceYearsAtThisCompany',
       'q_ExperienceYearsInCurrentRole', 'q_YearsSinceLastPromotion',
       'q_YearsWithCurrManager', 'c_Attrition', 'y_PerformanceRating'],
      dtype='object')

In [32]:
# Create All lists
AllColumns = ['c_EmpNumber', 'q_Age', 'c_Gender', 'c_EducationBackground',
       'c_MaritalStatus', 'c_EmpDepartment', 'c_EmpJobRole',
       'c_BusinessTravelFrequency', 'q_DistanceFromHome',
       'q_EmpEducationLevel', 'q_EmpEnvironmentSatisfaction',
       'q_EmpHourlyRate', 'q_EmpJobInvolvement', 'q_EmpJobLevel',
       'q_EmpJobSatisfaction', 'q_NumCompaniesWorked', 'c_OverTime',
       'q_EmpLastSalaryHikePercent', 'q_EmpRelationshipSatisfaction',
       'q_TotalWorkExperienceInYears', 'q_TrainingTimesLastYear',
       'q_EmpWorkLifeBalance', 'q_ExperienceYearsAtThisCompany',
       'q_ExperienceYearsInCurrentRole', 'q_YearsSinceLastPromotion',
       'q_YearsWithCurrManager', 'c_Attrition', 'y_PerformanceRating']
ConColList= ['q_Age','q_DistanceFromHome','q_EmpEducationLevel', 'q_EmpEnvironmentSatisfaction','q_EmpHourlyRate', 'q_EmpJobInvolvement', 'q_EmpJobLevel',
       'q_EmpJobSatisfaction', 'q_NumCompaniesWorked','q_EmpLastSalaryHikePercent', 'q_EmpRelationshipSatisfaction',
       'q_TotalWorkExperienceInYears', 'q_TrainingTimesLastYear','q_EmpWorkLifeBalance', 'q_ExperienceYearsAtThisCompany',
       'q_ExperienceYearsInCurrentRole', 'q_YearsSinceLastPromotion','q_YearsWithCurrManager']
CatColList=['c_EmpNumber', 'c_Gender', 'c_EducationBackground','c_MaritalStatus', 'c_EmpDepartment', 'c_EmpJobRole',
       'c_BusinessTravelFrequency','c_OverTime','c_Attrition']

In [33]:
# Instead of deleting created a new data frame without c_education_num
df2=df1[AllColumns] 
df2.head()

c_EmpNumber  q_Age c_Gender c_EducationBackground c_MaritalStatus  \
0    E1001000     32     Male             Marketing          Single   
1    E1001006     47     Male             Marketing          Single   
2    E1001007     40     Male         Life Sciences         Married   
3    E1001009     41     Male       Human Resources        Divorced   
4    E1001010     60     Male             Marketing          Single   

   c_EmpDepartment     c_EmpJobRole c_BusinessTravelFrequency  \
0            Sales  Sales Executive             Travel_Rarely   
1            Sales  Sales Executive             Travel_Rarely   
2            Sales  Sales Executive         Travel_Frequently   
3  Human Resources          Manager             Travel_Rarely   
4            Sales  Sales Executive             Travel_Rarely   

   q_DistanceFromHome  q_EmpEducationLevel  ...  \
0                  10                    3  ...   
1                  14                    4  ...   
2                   5                    4  ...   
3                  10                    4  ...   
4                  16                    4  ...   

   q_EmpRelationshipSatisfaction  q_TotalWorkExperienceInYears  \
0                              4                            10   
1                              4                            20   
2                              3                            20   
3                              2                            23   
4                              4                            10   

   q_TrainingTimesLastYear  q_EmpWorkLifeBalance  \
0                        2                     2   
1                        2                     3   
2                        2                     3   
3                        2                     2   
4                        1                     3   

   q_ExperienceYearsAtThisCompany  q_ExperienceYearsInCurrentRole  \
0                              10                               7   
1                               7                               7   
2                              18                              13   
3                              21                               6   
4                               2                               2   

  q_YearsSinceLastPromotion  q_YearsWithCurrManager  c_Attrition  \
0                         0                       8           No   
1                         1                       7           No   
2                         1                      12           No   
3                        12                       6           No   
4                         2                       2           No   

   y_PerformanceRating  
0                    3  
1                    3  
2                    4  
3                    3  
4                    3  

[5 rows x 28 columns]

# Encoding

In [34]:
# All Categorical variables to be Encoded starting from 0
from sklearn.preprocessing import LabelEncoder 
enc = LabelEncoder()
#df2.c_education = enc.fit_transform(df2.c_education)

for i in CatColList:
    df2[[i]] = enc.fit_transform(df1[[i]])

#Check if encoding is accurate
df2.head()

C:\Users\SANKET\anaconda3\lib\site-packages\sklearn\utils\validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)


c_EmpNumber  q_Age  c_Gender  c_EducationBackground  c_MaritalStatus  \
0            0     32         1                      2                2   
1            1     47         1                      2                2   
2            2     40         1                      1                1   
3            3     41         1                      0                0   
4            4     60         1                      2                2   

   c_EmpDepartment  c_EmpJobRole  c_BusinessTravelFrequency  \
0                5            13                          2   
1                5            13                          2   
2                5            13                          1   
3                3             8                          2   
4                5            13                          2   

   q_DistanceFromHome  q_EmpEducationLevel  ...  \
0                  10                    3  ...   
1                  14                    4  ...   
2                   5                    4  ...   
3                  10                    4  ...   
4                  16                    4  ...   

   q_EmpRelationshipSatisfaction  q_TotalWorkExperienceInYears  \
0                              4                            10   
1                              4                            20   
2                              3                            20   
3                              2                            23   
4                              4                            10   

   q_TrainingTimesLastYear  q_EmpWorkLifeBalance  \
0                        2                     2   
1                        2                     3   
2                        2                     3   
3                        2                     2   
4                        1                     3   

   q_ExperienceYearsAtThisCompany  q_ExperienceYearsInCurrentRole  \
0                              10                               7   
1                               7                               7   
2                              18                              13   
3                              21                               6   
4                               2                               2   

   q_YearsSinceLastPromotion  q_YearsWithCurrManager  c_Attrition  \
0                          0                       8            0   
1                          1                       7            0   
2                          1                      12            0   
3                         12                       6            0   
4                          2                       2            0   

   y_PerformanceRating  
0                    3  
1                    3  
2                    4  
3                    3  
4                    3  

[5 rows x 28 columns]

# Oulier Analysis
Identify Ouliers:
Outliers are data objects with characteristics that are considerably different than most of the other data objects in the data set. Identified using Sorting, min-max, clusturing, box plot, Std deviation, Distances scatter plots......

In [35]:
# Find outliers since z statitic can be applied only on continuous data
from scipy import stats
#if the outliers were present then, Remove the outliers and save the remaining data in a new data frame - df3
print('before :',df2.shape)
df3 = df2[(np.abs(stats.zscore(df1[ConColList]))<=3).all(axis=1)]
print('after :',df3.shape)

#Since about 2000 records are getting deleted, we choose pospone this deletion to next stage.

before : (1200, 28)
after : (1134, 28)


# Save 1134 records in the output file for next stage and exit

In [36]:
# This stage is over, so save and close the file
df2.to_csv('2_Employee_performance_withOutliers.csv',index=False)
df3.to_csv('2_Employee_performance_withOutOutliers.csv',index=False)

In [37]:
df3.to_csv('2_Adult_CE_withoutEncoding.csv',index=False)